### PR TITLE
dynet 2.1 (new formula)

### PR DESCRIPTION
**Pull request by ankane**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Adds DyNet, a neural network library from Carnegie Mellon University

### CODE REVIEW

File: Formula/dynet.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      system "make", "install"
+    end
+    pkgshare.install "examples"
+  end
+
+  test do
+    cp pkgshare/"examples/xor/train_xor.cc", testpath
+    system ENV.cxx, "-std=c++11", "train_xor.cc", "-I#{include}",
+                    "-L#{lib}", "-ldynet", "-o", "train_xor"
+    system "./train_xor"
```
Comment (SMillerDev) [inline Formula/dynet.rb]: Could we check for some output here to see if it's successful?

**Review by SMillerDev — COMMENTED** [review]
Thanks for the pull request, just a minor comment

**Comment by ankane** [comment]
Thanks for the feedback @SMillerDev, updated the PR. It looks like jobs on Jenkins are currently stuck.

**Review by chenrui333 — APPROVED** [review]

File: Formula/dynet.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      system "make", "install"
+    end
+    pkgshare.install "examples"
+  end
+
+  test do
+    cp pkgshare/"examples/xor/train_xor.cc", testpath
+    system ENV.cxx, "-std=c++11", "train_xor.cc", "-I#{include}",
+                    "-L#{lib}", "-ldynet", "-o", "train_xor"
+    _, stderr = Open3.capture3("./train_xor")
```
Comment (SMillerDev) [inline Formula/dynet.rb]: No need to use Open3, there's a `shell_output` command. You might need to redirect stderr to stdout though.

File: Formula/dynet.rb [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    assert_includes stderr, "E ="
```
Comment (SMillerDev) [inline Formula/dynet.rb]: Could you try to `assert_match` some more text? This doesn't seem like a lot.

**Review by SMillerDev — COMMENTED** [review]
Jenkins worked by the time I got to it. Some more feedback though but you're nearly there.

**Comment by ankane** [comment]
Updated the test. Here's the output of the script.

```txt
[dynet] random seed: 3776475602
[dynet] allocating memory: 512MB
[dynet] memory allocation done.
digraph G {
  rankdir=LR;
  nodesep=.05;
  N0 [label="v0 = parameters({8,2}) @ 0x7f9dec6000b8"];
  N1 [label="v1 = parameters({8}) @ 0x7f9dec6001d8"];
  N2 [label="v2 = parameters({1,8}) @ 0x7f9dec600318"];
  N3 [label="v3 = parameters({1}) @ 0x7f9dec600498"];
  N4 [label="v4 = constant({2})"];
  N5 [label="v5 = scalar_constant(0x7ffeee6567dc)"];
  N6 [label="v6 = v0 * v4"];
  N0 -> N6;
  N4 -> N6;
  N7 [label="v7 = v6 + v1"];
  N6 -> N7;
  N1 -> N7;
  N8 [label="v8 = tanh(v7)"];
  N7 -> N8;
  N9 [label="v9 = v2 * v8"];
  N2 -> N9;
  N8 -> N9;
  N10 [label="v10 = v9 + v3"];
  N9 -> N10;
  N3 -> N10;
  N11 [label="v11 = || v10 - v5 ||^2"];
  N10 -> N11;
  N5 -> N11;
}
E = 2.65871
E = 2.3155
E = 1.89155
E = 1.61723
E = 1.40677
E = 1.22902
E = 1.06623
E = 0.906574
E = 0.740014
E = 0.549427
E = 0.352105
E = 0.202604
E = 0.105996
E = 0.0511288
E = 0.023091
E = 0.00990864
E = 0.00409379
E = 0.00164662
E = 0.000650336
E = 0.000253782
E = 9.82667e-05
E = 3.78677e-05
E = 1.45432e-05
E = 5.57598e-06
E = 2.13541e-06
E = 8.16948e-07
E = 3.12782e-07
E = 1.19335e-07
E = 4.58043e-08
E = 1.74012e-08
```

The example uses a random seed, so the exact output is different every time. I think the best indication the test is working is a successful status code.

**Comment by SMillerDev** [comment]
Seems good to me. "Memory allocation" text is okay as a text check so good to go iyam.

**Comment by chenrui333** [comment]
Thanks @ankane for your contribution to Homebrew! :tada: 🥇 

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Comment by ankane** [comment]
Thanks @chenrui333 and @SMillerDev!

**Pull request closed without merging**